### PR TITLE
[Merged by Bors] - Add pod enrichment controller for injecting node addresses

### LINKED ISSUE #35: Annotate pods with node address
https://github.com/stackabletech/hdfs-operator/pull/175 (and others, like Kafka, will have the same problem later on) currently have an init job that reads the `Node' object to get the external IP. This embeds some annoying policy into the product container (such as which address to prefer), and requires the product Pod to have excessive privileges (you need a ClusterRoleBinding to be able to access Node objects).

I'd like to propose that we add a controller to the commons operator that pulls this information out into annotations on the Pod object (which we can then add a projected environment variable that reads).

The existing Pod.status.hostIP field is not sufficient, because:

- This typically prefers the internal IP rather than external
- Depending on cluster policy we'd like to be able to use hostnames here in the future (such as for Kerberos)

### PR DESCRIPTION
**Pull request by teozkr**
## Description

Fixes #35

## Review Checklist
- [ ] Code contains useful comments
- [ ] (Integration-)Test cases added (or not applicable)
- [ ] Documentation added (or not applicable)
- [ ] Changelog updated (or not applicable)
- [ ] Cargo.toml only contains references to git tags (not specific commits or branches)
- [ ] Helm chart can be installed and deployed operator works (or not applicable)

Once the review is done, comment `bors r+` (or `bors merge`) to merge. [Further information](https://bors.tech/documentation/getting-started/#reviewing-pull-requests)

### CODE REVIEW

File: rust/operator-binary/src/pod_enrichment_controller.rs [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    let mut annotations = BTreeMap::new();
+
+    let node_addr = node
+        .status
+        .iter()
+        .flat_map(|s| &s.addresses)
+        .flatten()
+        .filter_map(|addr| Some((NodeAddressType::from_str(&addr.type_).ok()?, &addr.address)))
+        .min_by_key(|(ty, _)| *ty)
+        .map(|(_, addr)| addr);
```
Comment (sbernauer) [inline rust/operator-binary/src/pod_enrichment_controller.rs:114]: very clever, i like!

**Review by sbernauer — APPROVED** [review]
hdfs go brrrrr

**Review by soenkeliebau — COMMENTED** [review]
Overall looks good to me, just one question (haven't tested it yet, sorry). 
Would this annotate all pods that are out there? I didn't find anywhere that this is restricted to something carrying a specific Stackable label or similar?

File: rust/operator-binary/src/pod_enrichment_controller.rs [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+const ANNOTATION_NODE_ADDRESS: &str = "enrichment.stackable.tech/node-address";
+
+struct Ctx {
+    client: stackable_operator::client::Client,
+}
+
+#[derive(Snafu, Debug, EnumDiscriminants)]
+#[strum_discriminants(derive(IntoStaticStr))]
+pub enum Error {
+    GetNode {
```
[the remark this answers is not in the document]
Comment (teozkr) [reply rust/operator-binary/src/pod_enrichment_controller.rs:30]: Good catch, forgot to write proper errors.

File: docs/modules/ROOT/pages/pod-enrichment.adoc [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+The Stackable Commons Operator automatically adds commonly used information to `Pod` objects, which
+would otherwise have to be inferred by traversing the Kubernetes object graph.
+
+== Node Address
+
+Annotation:: `enrichment.stackable.tech/node-address`
+
+The hostname or IP address of the `Node` that the `Pod` is scheduled to run on.
+Compared to `Pod.status.nodeIP`, this can also (but doesn't have to) be a hostname, and prefers
+externally routable addresses.
```
[the remark this answers is not in the document]
Comment (teozkr) [reply docs/modules/ROOT/pages/pod-enrichment.adoc:12]: We might add a separate `node-ip` at some point if something requires that specifically, but part of the point here is to abstract away the fact that not all clus†ers have meaningful (resolvable) hostnames for nodes.

**Comment by teozkr** [comment]
> Would this annotate all pods that are out there? I didn't find anywhere that this is restricted to something carrying a specific Stackable label or similar?

Yes, on the assumption that the annotation is namespaced anyway, so it shouldn't clash and anything that doesn't care can just ignore the annotation.

**Comment by soenkeliebau** [comment]
> > Would this annotate all pods that are out there? I didn't find anywhere that this is restricted to something carrying a specific Stackable label or similar?
> 
> Yes, on the assumption that the annotation is namespaced anyway, so it shouldn't clash and anything that doesn't care can just ignore the annotation.

Ok, fine for me. Is there anything to consider around this in an RBAC enabled environment? Would the operator keep throwing errors when trying to annotate pods it doesn't have the right to change or something like that?

**Comment by teozkr** [comment]
That's true, that'd happen for pods that it has permission to read but not patch…

**Comment by soenkeliebau** [comment]
> That's true, that'd happen for pods that it has permission to read but not patch…

We need to at some point look into having a stackableconfig configmap I think. We've had a couple of things come up that would be useful to configure centrally over the past few weeks. This is another one of those, being able to configure  labelselector to match pods that should get annotated would give us a mechanism to improve this, where needed.

File: rust/operator-binary/src/pod_enrichment_controller.rs [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    controller
+        .watches(
+            client.get_all_api::<Node>(),
+            ListParams::default(),
+            move |node| {
+                pods.state()
+                    .into_iter()
+                    .filter(move |pod| {
+                        pod.spec.as_ref().and_then(|s| s.node_name.as_deref())
+                            == node.metadata.name.as_deref()
```
[the remark this answers is not in the document]
Comment (teozkr) [reply rust/operator-binary/src/pod_enrichment_controller.rs:66]: Myeah, `ResourceExt` needs a bit of a revamp at some point: https://github.com/kube-rs/kube-rs/issues/634

**Comment by teozkr** [comment]
In that case I'd rather just hard-code a labelselector that triggers this..

**Comment by teozkr** [comment]
Added a filter to only enrich pods with the label `enrichment.stackable.tech/enabled: true`

**Review by soenkeliebau — APPROVED** [review]

**Pull request closed without merging**